src/zkbob/ZkBobPool.sol

```diff
@@ -242,6 +307,143 @@ contract ZkBobPool is EIP1967Admin, Ownable, Parameters, ZkBobAccounting {
+        emit CompleteDirectDepositBatch(_pool_index, _indices);
+    }
+
+    function directDeposit(address _fallbackUser, uint256 _amount, bytes memory _rawZkAddress) external {
+        _recordDirectDeposit(_fallbackUser, _amount, _rawZkAddress);
+    }
+
+    function onTokenTransfer(address _from, uint256 _value, bytes calldata _data) external returns (bool) {
+        require(msg.sender == token, "ZkBobPool: not a token caller");
+
+        (address fallbackUser, bytes memory rawZkAddress) = abi.decode(_data, (address, bytes));
+
+        _recordDirectDeposit(fallbackUser, _value, rawZkAddress);
+
+        return true;
+    }
+
+    function refundDirectDeposit(uint256 _index) external {
+        bool isOperator = operatorManager.isOperator(msg.sender);
+        DirectDeposit storage dd = directDeposits[_index];
+        require(dd.status == DirectDepositStatus.Pending, "ZkBobPool: direct deposit not pending");
+        require(
+            isOperator || dd.timestamp + directDepositTimeout < block.timestamp,
+            "ZkBobPool: direct deposit timeout not passed"
+        );
+        _refundDirectDeposit(_index, dd);
+    }
+
+    function refundDirectDeposit(uint256[] calldata _indices) external {
+        bool isOperator = operatorManager.isOperator(msg.sender);
+
+        for (uint256 i = 0; i < _indices.length; i++) {
+            DirectDeposit storage dd = directDeposits[_indices[i]];
+
+            if (dd.status == DirectDepositStatus.Pending) {
+                require(
+                    isOperator || dd.timestamp + directDepositTimeout < block.timestamp,
+                    "ZkBobPool: direct deposit timeout not passed"
+                );
+                _refundDirectDeposit(_indices[i], dd);
+            }
+        }
+    }
+
+    function _refundDirectDeposit(uint256 _index, DirectDeposit storage _dd) internal {
+        _dd.status = DirectDepositStatus.Refunded;
+
+        (address user, uint96 amount) = (_dd.user, _dd.amount);
+
+        IERC20(token).safeTransfer(user, amount);
+
+        emit RefundDirectDeposit(_index, user, amount);
+    }
+
+    function _recordDirectDeposit(address _fallbackUser, uint256 _amount, bytes memory _rawZkAddress) internal {
+        // TODO do something about remaining deposit dust (_amount % 1_000_000_000)
+        require(_fallbackUser != address(0), "ZkBobPool: fallback user is zero");
+
+        uint64 fee = directDepositFee;
+        uint64 depositAmount = uint64(_amount / TOKEN_DENOMINATOR);
+        require(depositAmount > fee, "ZkBobPool: direct deposit amount is too low");
+        unchecked {
+            depositAmount -= fee;
+        }
+
+        _checkDirectDepositLimits(msg.sender, depositAmount);
+
+        ZkAddress.ZkAddress memory zkAddress = ZkAddress.parseZkAddress(_rawZkAddress, uint24(pool_id));
+
+        DirectDeposit memory dd = DirectDeposit({
+            user: _fallbackUser,
+            amount: uint96(_amount),
+            deposit: depositAmount,
```

Comment:
should we keep or `amount` or `depositAmount` if one of this value can be recovered by using `fee`?

Reply:
We actually can't recover the exact amount, as some precision is lost at division `uint64(_amount / TOKEN_DENOMINATOR)`

Reply:
Either way, it will occupy the same amount of storage, so I don't think there is a notable overhead